src/Operations.jl

```diff
@@ -643,25 +643,55 @@ function download_artifacts(ctx::Context, pkgs::Vector{PackageSpec}; platform::A
     return download_artifacts(ctx, pkg_roots; platform=platform, verbose=verbose)
 end
 
+"""
+    load_pkg_hook(hook_path::String)
+
+Given a path to a hook file (which should end in a function definition) include
+it into an anonymous module and return that function.
+"""
+function load_pkg_hook(hook_path::String)
```

Comment:
I think this should be executed in a separate process to hide the side effects of executing the file (like if it loads dependencies) in the same way as `build.jl` does it.

Reply:
I'm conflicted on this; on the one hand, we have a natural serialization of platforms already defined, so communication is not a problem.  On the other, spinning up a subprocess to do this will add ~100ms * number of packages with hooks.  That's just on the cusp of being what I would consider to be expensive.

I've walked back on my stance that the hooks should be able to load things from their environment; they're now executed in whatever environment the `Pkg.add()` code itself is running in, so I think package authors will not be able to load things to effect things that much.  Even if they try to load a package, it would usually fail (because it may not have been installed yet, as this is happening mid-`Pkg.add()`) and if it succeeds, it will have come from the calling environment so it is probably the version you'd eventually load anyway.

Reply:
> On the other, spinning up a subprocess to do this will add ~100ms * number of packages with hooks. That's just on the cusp of being what I would consider to be expensive.

True, but there shouldn't be many packages doing this, it only happens at `install` time (where we have to download the package), etc. So will that really be significant?

Reply:
Not only at `install` time, but also things like `instantiate`, which will otherwise be much faster.  It will need to happen at `instantiate` time because we must check to see if each needed Artifact is actually installed or not.

Reply:
I just realized one reason the serialization idea won't work; if a Pkg hook needs to add a custom comparison strategy (e.g. CUDA doesn't just match `cuda+11` to `cuda+10`, it is a version cap, just like `os_version` or `libstdcxx_version`, so it needs to call `set_compare_strategy!()` on the platform that is passed through it).

Right now, we have a small library of alternate comparison strategies, but in general the comparison strategies are a `Dict{String,Function}`, and we don't want to serialize the `Function`'s as well, I don't think.  The comparison functions should be completely self-contained, and I'm fine with erroring if they're not, but I don't actually know if serializing functions across Julia sessions is supported.

Reply:
I still think it is wrong to include a bunch of files into the running julia process when running `Pkg.add`. Who knows what state the process now is in? What packages are loaded, what methods have been defined etc. Especially if you are running Pkg more as an API and not just as a user sitting in front of your terminal.